### DISCUSSION TITLE
GraphQL 在微服务架构中的实践 · /graphql-microservice

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#527] [Announcements]
https://draveness.me/graphql-microservice

**Comment by yetingsky** [comment]
请问下楼主你们是怎么监控 GraphQL 服务的呢？我们公司也有搭建类似 GraphQL 的服务，实行一段时间后碰到最大的问题是不好监控。之前的监控设置都是以一个 endpoint 来监控的，但现在 GraphQL 只有一个 endpoint，前端可以自由选择字段，这样的问题导致从 endpoint 的监控看不出啥（请求量，请求 500，请求响应时间），因为前段选择不同字段时，其实指标差距都是挺大的。所以想请教下楼主你们是怎么对 GraphQL 服务进行监控的？

**Comment by draveness** [comment]
> @yetingsky
> 请问下楼主你们是怎么监控 GraphQL 服务的呢？我们公司也有搭建类似 GraphQL 的服务，实行一段时间后碰到最大的问题是不好监控。之前的监控设置都是以一个 endpoint 来监控的，但现在 GraphQL 只有一个 endpoint，前端可以自由选择字段，这样的问题导致从 endpoint 的监控看不出啥（请求量，请求 500，请求响应时间），因为前段选择不同字段时，其实指标差距都是挺大的。所以想请教下楼主你们是怎么对 GraphQL 服务进行监控的？

这部分确实不大好做，不过 NewRelic 能解决大部分响应时间监控的问题，不同字段的监控目前也没有找到比较好用的开源项目，如果大部分字段都是通过 RPC 获取的，其实可以通过一些监控 RPC 的服务来解决一部分的监控问题，毕竟 GraphQL 只是一种对外提供 API 的方式。

**Comment by kdlan** [comment]
想请教一下你们关于错误处理方面是怎么进行的？

**Comment by kdlan** [comment]
还有 [Reconciling GraphQL and Thrift at Airbnb](https://medium.com/airbnb-engineering/reconciling-graphql-and-thrift-at-airbnb-a97e8d290712) 这篇不知道你们怎么看。通过 idl 自动生成graphql schema这点看起来很有吸引力。可以大大减少 graphql  网关的开发量

**Comment by draveness** [comment]
> @kdlan
> 想请教一下你们关于错误处理方面是怎么进行的？

1. 错误处理是指什么
2. 文章简单看了下，这种方式还不错，能够满足一些基本的需求，但是生成的 Schema 不能展现 GraphQL 的所有潜力和功能，如果对外暴露了其他接口形式 RESTful，可以尝试这种风格作为实验和补充，仅对外暴露一个 GraphQL 端点的服务还是实现一个完备的图比较好

**Comment by huangjimmy** [comment]
请问下你们用的哪一个GraphQL实现呢。我找了很多地方，都是规范，似乎没有一个参考实现啊。难道要自己从头实现吗？

**Comment by draveness** [comment]
> @huangjimmy
> 请问下你们用的哪一个GraphQL实现呢。我找了很多地方，都是规范，似乎没有一个参考实现啊。难道要自己从头实现吗？

我们后端有 Rails、Elixir 和 Golang 都有对应的实现，有的是官方的，有的是一些开发者自己做的开源项目

**Comment by kdlan** [comment]
> @Draveness
> > @kdlan
> > 想请教一下你们关于错误处理方面是怎么进行的？
> 
> 1. 错误处理是指什么
> 2. 文章简单看了下，这种方式还不错，能够满足一些基本的需求，但是生成的 Schema 不能展现 GraphQL 的所有潜力和功能，如果对外暴露了其他接口形式 RESTful，可以尝试这种风格作为实验和补充，仅对外暴露一个 GraphQL 端点的服务还是实现一个完备的图比较好

举个例子，比如好友列表上展示每个好友的最新动态。获取好友列表对应一个 resolver，每个好友的最新状态，可以利用 dataloader 来批量获取。假设这里获取好友的最新状态 rpc 失败了(网络异常、服务挂了等等)，那graphql返回的内容，是返回 

*  好友列表，并且状态提示错误
*  好友列表，状态为空，不提示错误
*  整个query返回错误

这里第1种相对比较准确，但是感觉实现起来并不方便。第二种和第三种都方便，但是第二种丢失了错误状态。第三种又过于粗暴。不知道你们针对这种情况是怎么处理的

**Comment by draveness** [comment]
> @kdlan
> > @Draveness
> > > @kdlan
> > > 想请教一下你们关于错误处理方面是怎么进行的？
> > 
> > 1. 错误处理是指什么
> > 2. 文章简单看了下，这种方式还不错，能够满足一些基本的需求，但是生成的 Schema 不能展现 GraphQL 的所有潜力和功能，如果对外暴露了其他接口形式 RESTful，可以尝试这种风格作为实验和补充，仅对外暴露一个 GraphQL 端点的服务还是实现一个完备的图比较好
> 
> 举个例子，比如好友列表上展示每个好友的最新动态。获取好友列表对应一个 resolver，每个好友的最新状态，可以利用 dataloader 来批量获取。假设这里获取好友的最新状态 rpc 失败了(网络异常、服务挂了等等)，那graphql返回的内容，是返回 
> 
> *  好友列表，并且状态提示错误
> *  好友列表，状态为空，不提示错误
> *  整个query返回错误
> 
> 这里第1种相对比较准确，但是感觉实现起来并不方便。第二种和第三种都方便，但是第二种丢失了错误状态。第三种又过于粗暴。不知道你们针对这种情况是怎么处理的

第一种其实就是标准的做法

**Comment by flaneur2020** [comment]
> 我们发现了微服务与 GraphQL 之间设计思路冲突的地方，也就是去中心化与中心化。

在技术选型时也遇到这个问题，我想我们将去中心化放在了优先位置。当时的考虑似乎主要是：

1. 上线单点问题：gateway 式的请求聚合入口可能要被全公司去修改，全公司一天不止几十次上线，联动 gateway 上线的话，发布风险会集中于单点
2. 开发过程连调：与端连调需要增加一个 gateway 层次，分层分仓库做 rpc 通信连调变得更难
3. 可用性衡量：不同部门希望按自己的接口计算可用性，如果是微服务垂直隔离的 RESTful 接口，这点很容易做到，而 gateway 本身的可用性可能影响其他部门，发布风险的问题可能会集中暴露在这里。

这几各问题感觉在 facebook 那种 单仓库 版本控制结合灰度发布机制下可能并不存在。

关于 GraphQL 本身， 个人印象也比较担心这几个问题：

- 结构依赖： GraphQL 的查询深入依赖 schema 的结构设计，如果前期设计 schema 不佳，需要迭代修改，会不会需要更多功夫？iOS / Android 如果对此类接口有直接依赖，可能两年下不掉。

这类迭代发布的问题在 react native 这种动态发布机制 + relay 这种高度整体设计的前端框架下可能也不存在。

而多仓库、多部门、微服务、非 relay 框架的环境里面，社区对 GraphQL 的工程经验感觉仍未十分充分，请教下楼主怎么看待这几个问题？

**Comment by draveness** [comment]
> @Fleurer
> > 我们发现了微服务与 GraphQL 之间设计思路冲突的地方，也就是去中心化与中心化。
> 
> 在技术选型时也遇到这个问题，我想我们将去中心化放在了优先位置。当时的考虑似乎主要是：
> 
> 1. 上线单点问题：gateway 式的请求聚合入口可能要被全公司去修改，全公司一天不止几十次上线，联动 gateway 上线的话，发布风险会集中于单点
> 2. 开发过程连调：与端连调需要增加一个 gateway 层次，分层分仓库做 rpc 通信连调变得更难
> 3. 可用性衡量：不同部门希望按自己的接口计算可用性，如果是微服务垂直隔离的 RESTful 接口，这点很容易做到，而 gateway 本身的可用性可能影响其他部门，发布风险的问题可能会集中暴露在这里。
> 
> 这几各问题感觉在 facebook 那种 单仓库 版本控制结合灰度发布机制下可能并不存在。
> 
> 关于 GraphQL 本身， 个人印象也比较担心这几个问题：
> 
> - 结构依赖： GraphQL 的查询深入依赖 schema 的结构设计，如果前期设计 schema 不佳，需要迭代修改，会不会需要更多功夫？iOS / Android 如果对此类接口有直接依赖，可能两年下不掉。
> 
> 这类迭代发布的问题在 react native 这种动态发布机制 + relay 这种高度整体设计的前端框架下可能也不存在。
> 
> 而多仓库、多部门、微服务、非 relay 框架的环境里面，社区对 GraphQL 的工程经验感觉仍未十分充分，请教下楼主怎么看待这几个问题？

1. 上线单点问题
    + 作为一个中心化的接口提供方式，遇到这种单点问题其实是无法避免的，我能想到的一种比较好的方式是将核心业务和其他业务分离，核心业务由于其特性会比较稳定，变动不会特别大，这两部分业务分别对其他服务的 RPC 请求进行编排或者协同，核心业务和其他业务分离后，可以提供多个 GraphQL 服务，或者在上层通过 Schema Stitching 的方式进行粘合。
2. 开发过程联调
    + 这个问题其实并不是 GraphQL 服务本身的问题，而是所有的微服务架构都需要解决的问题
3. 可用性衡量
    + 在内部，各个基础服务使用的其实都是 RPC 进行通信，最外层的几个服务会对内部服务提供的 RPC 进行编排，我们是使用 linkerd 加上了 Grafana 对所有内部的 gRPC 请求进行了监控，对于最外部的 GraphQL 服务可以使用常规的监控方式接入第三方或者自建监控相关的基础设施。
4. 结构依赖
    + 会肯定会，而且一定会花很多时间去做，前期很多工程师其实并不熟悉 Schema 的设计和规范，所以出现设计的问题已经要尽快推动依赖方修改相关的接口。

**Comment by gfreezy** [comment]
级联层级比较深的时候怎么解决性能问题

**Comment by draveness** [comment]
> @gfreezy
> 级联层级比较深的时候怎么解决性能问题

1. 限制级联层级；
2. 对每一个 Query 访问的节点数进行限制；
3. 对同一级别的查询进行合并查询，避免 N + 1 问题

**Comment by sirius1024** [comment]
GraphQL resolver的性能问题有没有最佳实践呢？

**Comment by wwulfric** [comment]
作者现在如何看待 graphql 的利弊？是便利居多还是麻烦居多

**Comment by jacktrane** [comment]
题主请问图是用什么画的,很好看

**Comment by draveness** [comment]
> @wwulfric
> 作者现在如何看待 graphql 的利弊？是便利居多还是麻烦居多

最近觉得 GraphQL 的接口应该由前端团队自行维护，后端团队维护的服务最好还是各自提供接口，这样团队之间能够尽量降低沟通成本，前端的同学也可以通过 GraphQL 批量获取资源，降低前端的加载时间，优化性能

**Comment by draveness** [comment]
> @jacktrane
> 题主请问图是用什么画的,很好看

看下结尾吧

**Comment by melonboy312** [comment]
题主，请问下你们是怎么对schema管理的呢？下面多个rpc的服务提供者，增加各种实体的时候，schema是都修改graphql服务中的schema吗还是有专门的管理工具和合并、校验工具呢？以及不同Type对应的RPC接口是怎么发现的呢？可以分享一下解决思路吗？

**Comment by draveness** [comment]
> @melonboy312
> 题主，请问下你们是怎么对schema管理的呢？下面多个rpc的服务提供者，增加各种实体的时候，schema是都修改graphql服务中的schema吗还是有专门的管理工具和合并、校验工具呢？以及不同Type对应的RPC接口是怎么发现的呢？可以分享一下解决思路吗？

各自服务修改各自的 Schema，没有工具去管理，每个服务提供的 RPC 接口其实就是 GraphQL 的 HTTP 接口

> GraphQL 前端服务总共包含两个核心组件，分别是 GraphQL Stitcher 和 Dispatcher，其中前者负责向各个 GraphQL 服务请求 IntrospectionQuery 并将获得的所有 Schema 粘合成一颗巨大的树；当客户端进行请求时，Graphql Dispatcher 会通过语法解析当前的请求，并将其中不同的字段以及子字段转换成树后转发给对应的服务。
>
> 在实现 GraphQL Stitcher 的过程中，需要格外注意不同服务之间类型冲突的情况，我们在实现的过程中并没有支持类型冲突以及跨服务资源的问题，而是采用了覆盖的方式，这其实有很大的问题，内部的 GraphQL 服务其实并不知道整个 Schema 中有哪些类型是已经被使用的，所以经常会造成服务之间的类型冲突，我们只有在发现时手动增加前缀来解决冲突。

**Comment by sandersyao** [comment]
如果服务端响应schema是否也应该根据用户的授权状况来响应？毕竟我不希望随便谁都能看遍整张图的结构。

**Comment by draveness** [comment]
> 如果服务端响应schema是否也应该根据用户的授权状况来响应？毕竟我不希望随便谁都能看遍整张图的结构。

这个东西你可以另外做吧，据我了解 GraphQL 对你说的这功能没支持

**Comment by codebdy** [comment]
Schema Stitching 是不是已经不大用了，开始转向apollo Fedretion了

**Reply by codebdy** [reply]
Federation
